simpledotorg/simple-server
# Add API docs and mock server for import API

We add an import API in Simple behind a feature flag. It shall currently be a mock server that only runs the validation on the request and return a success or failure. We also generate swagger documentation for this API.

**Story card:** [sc-10849](https://app.shortcut.com/simpledotorg/story/10849) and [sc-10850](https://app.shortcut.com/simpledotorg/story/10850)

## Because

Partner organizations in Bangladesh need a reference to build the clients that integrate with us. They have requested detailed API specifications as well as a mock server.

## This addresses

* Adds API with payload validations to the `/import` endpoint.
* Adds a not-currently-exhaustive integration test suite which mostly behaves as a smoke test for now. We are deferring making this exhaustive, covering all resource types, etc., to get the mock server out and unblock the partner orgs.
* Adds detailed and descriptive Swagger specs.

## Not addressed

* OAuth endpoints are not covered. Partner organizations can test against something like [this](https://oauth.tools/collection/1599045253169-GHF) while we build this into Simple.
* The mock import endpoint currently works without an `Authorization` header, but this should not be a problem for those needing this as a development reference.

## Test instructions

* Launch the Simple Server.
* Ensure http://localhost:3000/api-docs has a page called `Import` with the API spec.
* Add flipper flag called `imports_api`.
* Run this request using Postman/cURL whatever, adjust the payload to be anything in the spec:
```
PUT http://localhost:3000/api/v4/import
Content-Type: application/json; charset=UTF-8
HTTP_X_ORGANIZATION_ID: 220d5e23-dc13-47c6-9ef7-30c51b1d4456

{"resources":[{"resourceType":"Patient","meta":{"lastUpdated":"2019-08-24T14:15:22Z","createdAt":"2019-08-24T14:15:22Z"},"identifier":[{"value":"string"}],"active":true,"name":{"text":"string"},"telecom":[{"value":"string","use":"home"}],"gender":"male","birthDate":"2019-08-24","deceasedBoolean":true,"managingOrganization":[{"value":"string"}],"registrationOrganization":[{"value":"string"}],"address":[{"line":["string"],"city":"string","district":"string","state":"string","postalCode":"string"}]}]}
```

The above should give `202 Accepted` with no errors. Any other payload would give a `400 Bad Request`.
